authorization/http.go

```diff
@@ -44,3 +52,90 @@ func WithAuthorizers(authorizers map[string]rbac.Authorizer, permission rbac.Per
 		})
 	}
 }
+
+// WithEnforceTenantLabel returns a middleware that ensures that every query
+// has a tenant label enforced.
+func WithEnforceTenantLabel(label string) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		// Adapted from
+		// https://github.com/prometheus-community/prom-label-proxy/blob/952266db4e0b8ab66b690501e532eaef33300596/injectproxy/routes.go.
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			id, ok := authentication.GetTenantID(r.Context())
+			if !ok {
+				http.Error(w, "error finding tenant ID", http.StatusInternalServerError)
+				return
+			}
+
+			e := injectproxy.NewEnforcer([]*labels.Matcher{{
+				Name:  label,
+				Type:  labels.MatchEqual,
+				Value: id,
+			}}...)
+
+			// The `query` can come in the URL query string and/or the POST body.
+			// For this reason, we need to try to enforcing in both places.
+			// Note: a POST request may include some values in the URL query string
+			// and others in the body. If both locations include a `query`, then
+			// enforce in both places.
+			q, found1, err := enforceQueryValues(e, r.URL.Query())
+			if err != nil {
+				http.Error(w, fmt.Sprintf("could not enforce tenant label: %v", err), http.StatusInternalServerError)
+				return
+			}
+			r.URL.RawQuery = q
+
+			var found2 bool
+			// Enforce the query in the POST body if needed.
+			if r.Method == http.MethodPost {
+				if err := r.ParseForm(); err != nil {
+					http.Error(w, fmt.Sprintf("could not parse form: %v", err), http.StatusInternalServerError)
+					return
+				}
+				q, found2, err = enforceQueryValues(e, r.PostForm)
+				if err != nil {
+					http.Error(w, fmt.Sprintf("could not enforce tenant label: %v", err), http.StatusInternalServerError)
+					http.Error(w, "no query found", http.StatusInternalServerError)
+					return
+				}
+				// We are replacing request body, close previous one (ParseForm ensures it is read fully and not nil).
+				_ = r.Body.Close()
+				r.Body = ioutil.NopCloser(strings.NewReader(q))
+				r.ContentLength = int64(len(q))
+			}
+
+			// If no query was found, return early.
+			if !found1 && !found2 {
+				http.Error(w, "no query found", http.StatusBadRequest)
+				return
+			}
+
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+const queryParam = "query"
+
+// Adapted from
+// https://github.com/prometheus-community/prom-label-proxy/blob/952266db4e0b8ab66b690501e532eaef33300596/injectproxy/routes.go.
+func enforceQueryValues(e *injectproxy.Enforcer, v url.Values) (values string, noQuery bool, err error) {
```

Comment:
```suggestion
func enforceQueryValues(e *injectproxy.Enforcer, v url.Values) (values string, foundQuery (or queryPresent) bool, err error) {
```

This confused me at first that we are setting `noQuery` to `false` when there is "no query".

Reply:
this is taken verbatim from the prom-label-proxy, but this is really backwards. Let's fix that :+1: